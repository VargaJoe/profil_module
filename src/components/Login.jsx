import * as React                   from 'react';
import logo 						from '../images/logo.png';

export const Login = ({ formSubmit }) => {
    let nameInput, passwordInput;
    const onSubmit = e => {
        e.preventDefault();
        formSubmit(e, nameInput.value, passwordInput.value);
    };
    return (
        <div className="fix_block">
            <img src={logo} alt="site logo" />
            <div className="login_page">
                <h1>Login</h1>
                <form onSubmit={e => { onSubmit(e); }}>
                    <input type="text" placeholder="loginname" ref={el => nameInput = el} />
                    <input type="password" placeholder="password" ref={el => passwordInput = el} />
                    <button className="sn_btn">login</button>
                </form>
            </div>
        </div>
    );
};

export default Login; 