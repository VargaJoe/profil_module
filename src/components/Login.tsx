import * as React                   from 'react';
import { Link }                     from 'react-router-dom';
const logo = require('../images/logo.png');

interface Props {
    formSubmit: (email: string, pass: string) => void;
}

interface State {
    emailInput: string;
    passwordInput: string;
    error: boolean;
    emailValid: boolean;
    passwordValid: boolean;
}

export class Login extends React.Component<Props, State> {

    constructor(props: any) {
        super(props);
        this.state = {
            emailInput: '', 
            passwordInput: '',
            emailValid: true,
            passwordValid: true,
            error: false,
        },

        this.onSubmit               = this.onSubmit.bind(this);
        this.handleUserEmail        = this.handleUserEmail.bind(this);
        this.handleUserPassword     = this.handleUserPassword.bind(this);
        this.handleBlur             = this.handleBlur.bind(this);
    }
    
    public onSubmit = (e: any) => {
        e.preventDefault();
        this.props.formSubmit(this.state.emailInput, this.state.passwordInput);
    }

    public handleUserEmail = (e: any) => {
        const emailValid = e.target.value.match(/^[A-Za-z\s0-9\\]+$/);
        this.setState({
            emailInput: e.target.value,
        });
        if ( emailValid && e.target.value.length > 0) {
            this.setState({
                emailValid: true
            });
        } else {
            this.setState({
                error: true,
                emailValid: false
            });
        }
    }

    public handleUserPassword = (e: any) => {
        const regex = /^[A-Za-z\s0-9\\]+$/;
        const special = !e.target.value.match(/[!@#$%\^&*\+]/);
        let passwordValid = regex.test(e.target.value) && special ? true : false;
        this.setState({
            passwordInput: e.target.value,
        });
        // console.log(e.target.value);
        if ( passwordValid ) {
            this.setState({
                passwordValid: true
            });
            
        } else {
            this.setState({
                error: true,
                passwordValid: false
            });
        }
    }

    public handleBlur = () => {
        if (this.state.emailValid && this.state.passwordValid) {
            this.setState({
                error: false
            });
        }
    }

    public render () {
        console.log('render');
        return (
            <div className="fix_block">
                <Link to="/" >
                    <img src={logo} alt="site logo" />
                </Link>
                <div className="login_page">
                    <h1>Login</h1>
                    <form onSubmit={e => { this.onSubmit(e); }}>
                        <input 
                            type="text" 
                            required={true} 
                            placeholder="loginname" 
                            name="email"
                            data-testid="email"
                            onChange={this.handleUserEmail}
                            value={this.state.emailInput}
                            onBlur={this.handleBlur}
                        />
                        <input 
                            type="password" 
                            required={true} 
                            placeholder="password" 
                            name="password"
                            className="login__password"
                            data-testid="password"
                            onChange={this.handleUserPassword}
                            value={this.state.passwordInput}
                            onBlur={this.handleBlur}
                        />
                        <button 
                            className="sn_btn"
                            id="submitLoginForm"
                            disabled={this.state.passwordValid && this.state.emailValid ? false : true}
                        >
                            login
                        </button>
                        <Link to="/forgottenPassword" className="forgott_pass" >
                            Forgott password
                        </Link>
                        
                        { this.state.error ? 
                            ( 
                                <div className="error">
                                    Values atre not valid
                                </div>
                            )
                            :
                            ( '' )
                        }
                    </form>
                </div>
            </div>
        );
    }
}

export default Login; 