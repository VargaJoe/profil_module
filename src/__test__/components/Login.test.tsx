import * as React 						from 'react';
import Login 							from '../../components/Login';
import { 
	configure, 
    shallow,
    mount, 
    ReactWrapper } 						from 'enzyme';
import * as Adapter 					from 'enzyme-adapter-react-16';
import toJson 				      		from 'enzyme-to-json';

configure( {adapter: new Adapter()} );

const updateInput = (login: any, instance: any, newValue: any) => {
    const input = login.find(instance);
    input.simulate('change', {
        target: {value: newValue}
    });
    return login.find(instance);
};

const wrongEmails = ['somebody@', 'somebody@.', 'somebody@mail', 'somebodymail.com', 'somebody@mail.'];
const wrongPasswords = ['aaaaaa', 'aaaa@', '@@@@###', 'AAAAAA', 'AAAdfsdd', 'AAA1234', 'gdfdf1234', 'sd'];

describe('<Login /> rendering', () => {
    const props = {
        formSubmit: (emailInput: string, passwordInput: string) => {
            return 'sucsces';
        }
    };
    const login: ReactWrapper<any, any> = mount(<Login {...props} />); 

    Login.prototype.constructor(); 
    
	it('Contain one H1 element ', () => {
		expect(login.find('h1').length).toBe(1);
    });
    // test Snapshot 
	it('Match to snapshot', () => {
		expect(toJson(login)).toMatchSnapshot();
    });
     
    it('Income values are same as in state ', () => {
        const email = updateInput(login, '[data-testid="email"]', 'edgar@gmail.com');
        const password = updateInput(login, '[data-testid="password"]', 'abAB12');
        expect(email.props().value).toBe('edgar@gmail.com');
        expect(password.props().value).toBe('abAB12');
    }); 
    
    it('Email validation ', () => {
        for (let index = 0; index < wrongEmails.length; index++) {
            const email = updateInput(login, '[data-testid="email"]', wrongEmails[index]);
            expect(login.state().emailValid).toBe(false);
        }
    }); 
    
    it('Password validation ', () => {
        for (let index = 0; index < wrongPasswords.length; index++) {
            const password = updateInput(login, '[data-testid="password"]', wrongPasswords[index]);
            expect(login.state().passwordValid).toBe(false);
        }
    }); 

    it('Password change calls validation functions', () => {
        jest.spyOn(Login.prototype, 'handleBlur'); 
        jest.spyOn(Login.prototype, 'handleUserPassword'); 
        
        const input = login.find('[data-testid="password"]');
        input.simulate('change', {
            target: {value: 'abAB12'}
        });
        expect(Login.prototype.handleBlur.call.length).toBe(1);
        expect(Login.prototype.handleUserPassword.call.length).toBe(1);
    }); 

    it('Email change calls validation functions', () => {
        let spyBlur = jest.spyOn(Login.prototype, 'handleBlur'); 
        let spyEmail = jest.spyOn(Login.prototype, 'handleUserEmail'); 
        
        const input = login.find('[data-testid="email"]');
        Login.prototype.handleBlur();
        let event = {
            target: {value: 'abAB12'}
        };
        Login.prototype.handleUserEmail(event);
        input.simulate('change');
        expect(spyBlur).toHaveBeenCalledTimes(1);
        expect(spyEmail).toHaveBeenCalledTimes(1);
    }); 

    it('Password is called without errors', () => {
        let spyPass = jest.spyOn(Login.prototype, 'handleUserPassword'); 
        
        let input = login.find('[data-testid="password"]');

        let event = {
            target: {value: 'abAB12'}
        };
        Login.prototype.handleUserPassword(event);
        input.simulate('change');
        expect(spyPass).toHaveBeenCalledWith(event);
        expect(spyPass).toHaveBeenCalledTimes(1);
    }); 

    it('onSubmit is called ', () => {
        let spyPass = jest.spyOn(Login.prototype, 'onSubmit'); 
        let event = {
            target: {value: 'nothing'}
        };
        Login.prototype.onSubmit();
        expect(spyPass).toHaveBeenCalledTimes(1);
    }); 
    
    it('Error message ', () => {
        login.setState({ error: true });
        const errorMsg = login.find('.error');
        expect(errorMsg.length).toBe(1);
        login.setState({ error: false });
    }); 
});
