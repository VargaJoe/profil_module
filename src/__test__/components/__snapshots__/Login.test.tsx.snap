// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> rendering Match to snapshot 1`] = `
<Login
  formSubmit={[Function]}
>
  <div
    className="fix_block"
  >
    <img
      alt="site logo"
      src="logo.png"
    />
    <div
      className="login_page"
    >
      <h1>
        Login
      </h1>
      <form
        onSubmit={[Function]}
      >
        <input
          data-testid="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="loginname"
          required={true}
          type="text"
          value=""
        />
        <input
          className="login__password"
          data-testid="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="password"
          required={true}
          type="password"
          value=""
        />
        <button
          className="sn_btn"
          disabled={false}
          id="submitLoginForm"
        >
          login
        </button>
      </form>
    </div>
  </div>
</Login>
`;
